### PR TITLE
mgr/ansible: Integrate mgr/ansible/tox into mgr/tox

### PR DESCRIPTION
**Pull request by sebastian-philipp**
`orchestrator.PlacementSpec` sometimes had a `nodes` member
and sometimes had a `hosts` member. Unified it to `hosts`

Reason to make this switch is to be able to do proper refactorings
in the orchestrators.

- [x] **Blocked by #32236** (merged)
- [ ] will conflict with #32272

Signed-off-by: Sebastian Wagner <sebastian.wagner@suse.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by liewegas** [comment]
needs rebase

**Comment by sebastian-philipp** [comment]
rebased

File: src/pybind/mgr/ansible/ansible_runner_svc.py [hunk 1]
```diff
@@ -8,7 +8,7 @@
 import collections
 import logging
 try:
-    from typing import Optional
+    from typing import Optional, Dict, Any, List, Set
```
Comment (tchaikov) [inline src/pybind/mgr/ansible/ansible_runner_svc.py]: could you move these mypy related changes into another commit?

File: src/pybind/mgr/orchestrator.py [hunk 2]
```diff
@@ -1042,12 +1050,15 @@ class PlacementSpec(object):
     """
     For APIs that need to specify a node subset
     """
-    def __init__(self, label=None, nodes=None):
+    def __init__(self,
+                 label=None,  # type: Optional[str]
+                 hosts=None  # type: Optional[List[str]]
```
Comment (tchaikov) [inline src/pybind/mgr/orchestrator.py]: could you move s/nodes/hosts/ related changes into its own commit?

Comment (sebastian-philipp) [reply src/pybind/mgr/orchestrator.py]: You mean, I should first unify the mess of (nodes,hosts) to nodes and then rename everything back to hosts?

File: src/pybind/mgr/ansible/ansible_runner_svc.py (hunk 1 above)
Comment (sebastian-philipp) [reply src/pybind/mgr/ansible/ansible_runner_svc.py]: That would be artificial. I did all of those fixes in one go:

1. remove `mgr/ansible/tox.ini` + cmake integration
2. add ansible to `mgr/tox.ini`
3. Fix all issues that did pop up when doing so.

File: src/pybind/mgr/orchestrator.py (hunk 2 above)
Comment (tchaikov) [reply src/pybind/mgr/orchestrator.py]: why would you change them to nodes first?

File: src/pybind/mgr/ansible/ansible_runner_svc.py (hunk 1 above)
Comment (tchaikov) [reply src/pybind/mgr/ansible/ansible_runner_svc.py]: i believe atomic changes would help reviewers to understand them better.

File: src/pybind/mgr/orchestrator.py (hunk 2 above)
Comment (sebastian-philipp) [reply src/pybind/mgr/orchestrator.py]: exactly. that's why I converged to `hosts`

File: src/pybind/mgr/ansible/ansible_runner_svc.py (hunk 1 above)
Comment (sebastian-philipp) [reply src/pybind/mgr/ansible/ansible_runner_svc.py]: Actually this is one big atomic change. Either you don't check ansible with mypy, then you don't need to change anything. Or, you do check ansible with mypy and then you'll need to make all of these changes.

File: src/pybind/mgr/orchestrator.py (hunk 2 above)
Comment (tchaikov) [reply src/pybind/mgr/orchestrator.py]: > I should first unify the mess of having both (nodes and hosts) to nodes and then rename everything back to hosts?

@sebastian-philipp i am confused. let me rephrase: i don't think you should first unify the mess of having both (nodes and hosts) to nodes and then rename everything back to hosts. instead, i think a commit which just do PlacementSpec  and s/nodes/hosts/ change would be good enough.

Comment (sebastian-philipp) [reply src/pybind/mgr/orchestrator.py]: :+1:

File: src/pybind/mgr/orchestrator.py [hunk 3]
```diff
@@ -34,7 +34,15 @@
 logger = logging.getLogger(__name__)
 
 
-HostSpec = namedtuple('HostSpec', ['hostname', 'network', 'name'])
+class HostSpec(namedtuple('HostSpec', ['hostname', 'network', 'name'])):
```
Comment (tchaikov) [inline src/pybind/mgr/orchestrator.py]: could you move this change to another commit?

File: src/pybind/mgr/orchestrator.py [hunk 4]
```diff
@@ -1173,12 +1181,12 @@ class StatelessServiceSpec(object):
     # start the services.
 
     def __init__(self, name, placement=None, count=None):
-        self.placement = PlacementSpec() if placement is None else placement
+        self.placement = PlacementSpec() if placement is None else placement  # type: PlacementSpec
```
Comment (tchaikov) [inline src/pybind/mgr/orchestrator.py]: ditto.

File: src/pybind/mgr/orchestrator.py [hunk 5]
```diff
@@ -1236,7 +1244,7 @@ def __init__(self,
 
         #: List of hosts where RGWs should run. Not for Rook.
         if hosts:
-            self.placement.hosts = hosts
+            self.placement = PlacementSpec(hosts=hosts)
```
Comment (tchaikov) [inline src/pybind/mgr/orchestrator.py]: ditto.

File: src/pybind/mgr/ssh/tests/test_ssh.py [hunk 6]
```diff
@@ -115,7 +115,7 @@ def test_create_osds(self, _send_command, _get_connection, ssh_module):
     @mock.patch("ssh.module.SSHOrchestrator._get_connection")
     def test_mds(self, _send_command, _get_connection, ssh_module):
         with self._with_host(ssh_module, 'test'):
-            ps = PlacementSpec(nodes=['test'])
+            ps = PlacementSpec(hosts=['test'])
```
Comment (tchaikov) [inline src/pybind/mgr/ssh/tests/test_ssh.py]: ditto.

File: src/pybind/mgr/orchestrator.py (hunk 3 above)
Comment (sebastian-philipp) [reply src/pybind/mgr/orchestrator.py]: Eh, but this change is also required by mypy!?

File: src/pybind/mgr/orchestrator.py (hunk 5 above)
Comment (sebastian-philipp) [reply src/pybind/mgr/orchestrator.py]: ditto

File: src/pybind/mgr/orchestrator.py (hunk 3 above)
Comment (tchaikov) [reply src/pybind/mgr/orchestrator.py]: ahh, i didn't realize this. because mypy needs a type with the `__str__` magic method?

Comment (sebastian-philipp) [reply src/pybind/mgr/orchestrator.py]: done

File: src/pybind/mgr/orchestrator.py (hunk 4 above)
Comment (sebastian-philipp) [reply src/pybind/mgr/orchestrator.py]: done

File: src/pybind/mgr/ssh/tests/test_ssh.py (hunk 6 above)
Comment (sebastian-philipp) [reply src/pybind/mgr/ssh/tests/test_ssh.py]: done

**Comment by sebastian-philipp** [comment]
@tchaikov I've split this PR into multiple commits. Just be aware that everything only makes sense as one big entity, as the code only works at the end of the PR.

**Comment by sebastian-philipp** [comment]
rebased.

**Comment by sebastian-philipp** [comment]
```
2019-12-17T15:01:51.085 INFO:teuthology.orchestra.run.smithi101.stderr:  File "/usr/share/ceph/mgr/mgr_module.py", line 1064, in _handle_command
2019-12-17T15:01:51.085 INFO:teuthology.orchestra.run.smithi101.stderr:    return CLICommand.COMMANDS[cmd['prefix']].call(self, cmd, inbuf)
2019-12-17T15:01:51.086 INFO:teuthology.orchestra.run.smithi101.stderr:  File "/usr/share/ceph/mgr/mgr_module.py", line 304, in call
2019-12-17T15:01:51.086 INFO:teuthology.orchestra.run.smithi101.stderr:    return self.func(mgr, **kwargs)
2019-12-17T15:01:51.086 INFO:teuthology.orchestra.run.smithi101.stderr:  File "/usr/share/ceph/mgr/orchestrator.py", line 140, in wrapper
2019-12-17T15:01:51.086 INFO:teuthology.orchestra.run.smithi101.stderr:    return func(*args, **kwargs)
2019-12-17T15:01:51.087 INFO:teuthology.orchestra.run.smithi101.stderr:  File "/usr/share/ceph/mgr/orchestrator_cli/module.py", line 608, in _update_mons
2019-12-17T15:01:51.087 INFO:teuthology.orchestra.run.smithi101.stderr:    placement = orchestrator.PlacementSpec(label=label, count=num, nodes=hosts)
2019-12-17T15:01:51.087 INFO:teuthology.orchestra.run.smithi101.stderr:TypeError: __init__() got an unexpected keyword argument 'nodes'
2019-12-17T15:01:51.087 INFO:teuthology.orchestra.run.smithi101.stderr:) v1 ==== 612+0+0 (unknown 1541455595 0 0) 0x7fcba00009b0 con 0x7fcb98015130
2019-12-17T15:01:51.090 INFO:teuthology.orchestra.run.smithi101.stderr:Error EINVAL: Traceback (most recent call last):
2019-12-17T15:01:51.090 INFO:teuthology.orchestra.run.smithi101.stderr:  File "/usr/share/ceph/mgr/mgr_module.py", line 1064, in _handle_command
2019-12-17T15:01:51.090 INFO:teuthology.orchestra.run.smithi101.stderr:    return CLICommand.COMMANDS[cmd['prefix']].call(self, cmd, inbuf)
2019-12-17T15:01:51.090 INFO:teuthology.orchestra.run.smithi101.stderr:  File "/usr/share/ceph/mgr/mgr_module.py", line 304, in call
2019-12-17T15:01:51.091 INFO:teuthology.orchestra.run.smithi101.stderr:    return self.func(mgr, **kwargs)
2019-12-17T15:01:51.091 INFO:teuthology.orchestra.run.smithi101.stderr:  File "/usr/share/ceph/mgr/orchestrator.py", line 140, in wrapper
2019-12-17T15:01:51.091 INFO:teuthology.orchestra.run.smithi101.stderr:    return func(*args, **kwargs)
2019-12-17T15:01:51.091 INFO:teuthology.orchestra.run.smithi101.stderr:  File "/usr/share/ceph/mgr/orchestrator_cli/module.py", line 608, in _update_mons
2019-12-17T15:01:51.091 INFO:teuthology.orchestra.run.smithi101.stderr:    placement = orchestrator.PlacementSpec(label=label, count=num, nodes=hosts)
2019-12-17T15:01:51.091 INFO:teuthology.orchestra.run.smithi101.stderr:TypeError: __init__() got an unexpected keyword argument 'nodes'
```

**Comment by sebastian-philipp** [comment]
Fixed.
